src/channel/queuedChannel.ts

```diff
@@ -86,16 +78,46 @@ export class QueuedChannel<T> implements OutputChannel<T> {
         this.logger.debug(`Queue length: ${length}`);
 
         for (const message of this.queue.all()) {
-            this.pending = this.pending.then(
-                () => this.channel
-                    .publish(message)
-                    .then(this.dequeue.bind(this)),
-            );
+            this.pending = this.chainNext(this.pending, message);
         }
 
         return this.pending;
     }
 
+    private async chainNext(previous: Promise<void>, message: T, enqueue = false): Promise<void> {
+        if (enqueue) {
+            this.enqueue(message);
+        }
+
+        try {
+            await previous;
+        } catch (error) {
+            if (error instanceof MessageDeliveryError && error.retryable) {
+                // If the previous message failed to deliver, requeue all messages
+                // including the current one that was just enqueued
+                return this.requeue();
+            }
+
+            throw error;
+        }
+
+        try {
+            const result = await this.channel.publish(message);
+
+            this.dequeue();
+
+            return result;
+        } catch (error) {
+            if (!(error instanceof MessageDeliveryError) || !error.retryable) {
+                // Discard the message if it's non-retryable so that the next message
+                // in the queue can be processed
+                this.dequeue();
+            }
+
+            throw error;
+        }
```

Comment:
Scenario:
- Publishing a message throws a non-retriable error
- That message will be dequeued, but the logic of the second `catch`
- The error will be rethrown
- The following message will fail when awaiting `previous` because the error was rethrown
- The error is not retriable so it will be rethrown again the error. The queue won't resume.

So, any non-retriable error will be removed from the queue, but the queue will also be stopped. To completely flush the queue, one page refresh for each broken message would be needed.